### PR TITLE
Pv print

### PR DESCRIPTION
**Pull request by Akababa**
Prints to UCI iff the PV changed.
image

Code review welcome, I didn't want to touch too much.

### CODE REVIEW

**Comment by killerducky** [comment]
`UCTNode * top_child = nullptr;`  UCTNode is the most memory sensitive object, we could create up to 40 million of these (more when we add a UCI option to increase it). I don't think it's worth it to output an informational statistic to add 160MB of memory usage.

You're counting spaces in the pvstring? I think we should do better than that. Also calling getpv is relatively expensive since it locks every node as it goes down. Granted this isn't called that often.

I just now noticed you have #412 also, this is confusing...

See #364 which adds a recursive count to play_simulation. I think that method is better than this one, less memory.

**Comment by Akababa** [comment]
Good point about the memory usage; I hadn't thought of that.

I think it's fine to count spaces as the "best" way to get depth would probably be to have getpv() return it, but getpv returns a string, so I feel it's easiest and most consistent for now.

Also I'm not calling getpv any more than necessary (i.e. once per line of output), although the entire method is unnecessary with my 160MB top_child memoization. With that being said, it should be possible to keep the top_child always at the top of m_children with an atomic swap operation, removing the need for the pointer.

Speaking of memory, it looks like m_virtual_loss, m_init_eval, m_status and possibly m_is_expanding and m_has_children are extraneous? I could try to test these in another PR...

File: src/UCTNode.cpp [hunk 1]
```diff
@@ -282,7 +282,7 @@ void UCTNode::accumulate_eval(float eval) {
     atomic_add(m_whiteevals, (double)eval);
 }
 
-UCTNode* UCTNode::uct_select_child(Color color, bool is_root) {
+std::pair<UCTNode*, std::pair<bool,bool>> UCTNode::uct_select_child(Color color, bool is_root) {
```
Comment (gcp) [inline src/UCTNode.cpp]: Why not use a helper struct, or at least a tuple?

File: src/UCTNode.cpp [hunk 2]
```diff
@@ -315,8 +315,10 @@ UCTNode* UCTNode::uct_select_child(Color color, bool is_root) {
     // Estimated eval for unknown nodes = original parent NN eval - reduction
     // Or curent parent eval - reduction if dynamic_eval is enabled.
     auto fpu_eval = (cfg_fpu_dynamic_eval ? get_eval(color) : net_eval) - fpu_reduction;
+    auto bestit = m_children.begin();
 
-    for (const auto& child : m_children) {
+    for (auto it = m_children.begin(); it != m_children.end(); it++) {
```
Comment (gcp) [inline src/UCTNode.cpp]: You don't need to change this to a manual loop just to have a pointer to the best child. In fact, it's not clear to me why you insist on having to have an iterator (which is really a pointer)?

File: src/UCTNode.cpp [hunk 3]
```diff
@@ -334,11 +336,19 @@ UCTNode* UCTNode::uct_select_child(Color color, bool is_root) {
...
             best_value = value;
             best = child.get();
+            bestit = it;
         }
     }
 
+    bool pv_changed = (parentvisits == 0 ||
+        (bestit != m_children.begin() && best->get_visits() >= m_children.front()->get_visits()));
+    if (pv_changed) {
+        std::iter_swap(bestit, m_children.begin());
```
Comment (gcp) [inline src/UCTNode.cpp]: Can just use swap() here with pointers.

**Comment by gcp** [comment]
>Speaking of memory, it looks like m_virtual_loss, m_init_eval, m_status and possibly m_is_expanding and m_has_children are extraneous?

m_has_children is extraneous and was removed upstream.

m_virtual_loss is extraneous, but if you get rid of it, obtaining the "true" value of a node becomes impossible during smp search, because it requires you to know the calling stack of all other threads.

**Comment by Akababa** [comment]
> I would strongly consider only doing the "has the PV changed" check every x root iterations and simply doing a comparison on a get_pv call. Much simpler code, just as useful, should have no performance impact.

Ouch, yeah, that's probably right. Should I still bother with keeping the top child at the front or just let getpv do its thing...

**Comment by evalon32** [comment]
Could we do it once a second or such? UCI actually says to send infos "whenever one of the info has changed," but that can't be meant literally, because sending `info nodes` every time it changes... no thanks.

**Comment by Akababa** [comment]
Removed a ton of unnecessary stuff thanks to @gcp and replaced it with a simple check every 50 sims.

**Comment by gcp** [comment]
>Which upstream branch is that?

https://github.com/gcp/leela-zero/tree/next

**Comment by gcp** [comment]
Here's an example why things get tricky without split virtual_loss accounting: https://github.com/glinscott/leela-chess/pull/364#issuecomment-382640546

File: src/UCTNode.cpp [hunk 4]
```diff
@@ -316,7 +316,7 @@ UCTNode* UCTNode::uct_select_child(Color color, bool is_root) {
     // Or curent parent eval - reduction if dynamic_eval is enabled.
     auto fpu_eval = (cfg_fpu_dynamic_eval ? get_eval(color) : net_eval) - fpu_reduction;
 
-    for (const auto& child : m_children) {
+    for (const auto& child: m_children) {
```
Comment (killerducky) [inline src/UCTNode.cpp]: nit: can you put this back?

File: src/UCTSearch.h [hunk 5]
```diff
@@ -84,6 +84,8 @@ class UCTSearch {
     SearchResult play_simulation(BoardHistory& bh, UCTNode* const node);
 
 private:
+    std::string last_pv;
+    int last_depth;
```
Comment (killerducky) [inline src/UCTSearch.h]: Please move these down to be with all the other member variables, and follow the m_member_var naming convention (a few others have crept in, we can fix those later but let's not add more).

File: src/UCTSearch.cpp [hunk 6]
```diff
@@ -411,17 +415,22 @@ Move UCTSearch::think(BoardHistory&& new_bh) {
 
     bool keeprunning = true;
     int last_update = 0;
+    last_pv = "";
+    last_depth = 0;
+
+    if (m_nodes > 0) dump_analysis(Time.elapsed(), false); // so we don't say nothing
```
[the remark this answers is not in the document]
Comment (Akababa) [reply src/UCTSearch.cpp]: It isn't, thanks.

File: src/UCTSearch.h (hunk 5 above)
Comment (Akababa) [reply src/UCTSearch.h]: Thanks, done

**Comment by Akababa** [comment]
@killerducky This is relatively simple, so can it be merged before another conflict arises? Thanks.

**Comment by killerducky** [comment]
@mooskagh what do you think about this? Would it be possible to implement in the TF version?

**Comment by mooskagh** [comment]
Implement what exactly, PV update? Yes, it's possible, currently it only updates when the first move changes.
It's also possible to track PV change without periodic polling, store PV as vector<Node, N>, and track when some PV node sibling's N becomes larger.
Is seeing changes in PV tail actually useful/interesting?

**Comment by Akababa** [comment]
@killerducky Sure, that makes sense. Now that I can compile lc0 I was planning to take a closer look there anyway, so feel free to close this and I'll PR them together.

**Comment by Akababa** [comment]
@killerducky Sure, that makes sense. Now that I can compile lc0 I was planning to take a closer look there anyway, so feel free to close this and I'll PR them together.

**Pull request closed without merging**